h2o-core/src/main/java/water/Scope.java

```diff
@@ -20,56 +24,88 @@
  */
 
 public class Scope {
+  
+  private static final Logger log = LoggerFactory.getLogger(Scope.class);
+  
   // Thread-based Key lifetime tracking
-  static private final ThreadLocal<Scope> _scope = new ThreadLocal<Scope>() {
+  private static final ThreadLocal<Scope> _scope = new ThreadLocal<Scope>() {
     @Override protected Scope initialValue() { return new Scope(); }
   };
-  private final Stack<HashSet<Key>> _keys = new Stack<>();
+  private final Stack<Set<Key>> _keys = new Stack<>();
+  
+  private final Stack<Set<Key>> _protectedKeys = new Stack<>();
   
   /** debugging purpose */
-  static public Scope current() {
+  public static Scope current() {
     return _scope.get();
   }
-
+  
+  /** for testing purpose */
+  public static int currentLevel() {
+    Scope scope = current();
+    assert scope._keys.size() == scope._protectedKeys.size();
+    return scope._keys.size();
+  }
+  
+  /** for testing purpose */
+  public static void reset() {
+    Scope scope = _scope.get();
+    scope._keys.clear();
+    scope._protectedKeys.clear();
+  }
+  
   /** Enter a new Scope */
-  static public void enter() { _scope.get()._keys.push(new HashSet<Key>()); }
+  public static void enter() { 
+    Scope scope = _scope.get();
+    scope._keys.push(new HashSet<>()); 
+    Set<Key> outerProtected = scope._protectedKeys.empty() ? Collections.emptySet() : scope._protectedKeys.peek(); 
+    scope._protectedKeys.push(new HashSet<>(outerProtected)); //inherit protected keys from outer scope
+  }
 
-  /** Exit the inner-most Scope, remove all Keys created since the matching
+  /** Exit the innermost Scope, remove all Keys created since the matching
    *  enter call except for the listed Keys.
    *  @return Returns the list of kept keys. */
-  static public Key[] exit(Key... keep) {
-    List<Key> keylist = new ArrayList<>();
-    if( keep != null )
-      for( Key k : keep ) if (k != null) keylist.add(k);
-    Object[] arrkeep = keylist.toArray();
+  public static Key[] exit(Key... keep) {
+    Scope scope = _scope.get();
+    assert !(scope._keys.empty() || scope._protectedKeys.empty()): "Scope in inconsistent state: Scope.exit() called without a matching Scope.enter()";
+    Set<Key> keepKeys = new HashSet<>();
+    if (keep != null) {
+      for (Key k : keep) {
+        if (k != null) keepKeys.add(k);
+      }
+    }
+    keepKeys.addAll(scope._protectedKeys.pop());
```

Comment:
instead of checking protected keys at `exit`, I could also directly prevent them from being tracked in `track_impl` as this works also for implicit calls to `track_internal`: not sure this will be significantly more efficient thoough, as we're still doing a lookup for each tracked key on exit.